### DISCUSSION TITLE
【QUESTION】Generate and verify proof without Bonsai ?

### DISCUSSION THREAD

**Discussion by godeamon** [risc0/risc0#1288] [Q&A] [answered]
I would like to ask if you can generate proof and verify without using Bonsai and only use the code of the current code base. In other words, can you use the current open source code to complete the overall process of ZKP from setup, generating proof, and verify proof? Because I know that bonsai is not open source.

**Comment by it09** [comment]
You can use zkVM , which is open source

**Comment by pdg744** [comment, marked as answer]
Yes -- the open-source code includes everything you need!

This includes: 
- tooling for deterministically compiling Rust to RISC-V
- tooling for proving RISC-V binaries of [arbitrary size](https://www.risczero.com/blog/continuations)
- tooling for *aggregating* and *[composing](https://www.risczero.com/blog/proof-composition)* proofs into a single succinct proof
- tooling for verifying proofs, off-chain or on-chain. 

If you are proving without Bonsai, proofs for large programs are generated sequentially on your machine. 
If you are proving with Bonsai, proofs for large programs are generated in parallel on a scalable cluster of machines.
